Sources/StreamChat/Controllers/BackgroundListDatabaseObserver.swift

```diff
@@ -0,0 +1,297 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+import Foundation
+
+class ListDatabaseObserverWrapper<Item, DTO: NSManagedObject> {
+    private var foreground: ListDatabaseObserver<Item, DTO>?
+    private var background: BackgroundListDatabaseObserver<Item, DTO>?
+    let isBackground: Bool
+
+    var items: LazyCachedMapCollection<Item> {
+        if isBackground, let background = background {
+            return background.items
+        } else if let foreground = foreground {
+            return foreground.items
+        } else {
+            log.assertionFailure("Should have foreground or background observer")
+            return []
+        }
+    }
+
+    /// Called with the aggregated changes after the internal `NSFetchResultsController` calls `controllerWillChangeContent`
+    /// on its delegate.
+    var onWillChange: (() -> Void)? {
+        didSet {
+            if isBackground {
+                background?.onWillChange = onWillChange
+            } else {
+                foreground?.onWillChange = onWillChange
+            }
+        }
```

Comment:
All these ifs can be avoided if we create a `ListDatabaseObserving` protocol since the background and the regular observer have the same interface. Then, we would only need to do the if once, when creating the observer.